components/keys/src/types.rs

```diff
@@ -211,14 +212,108 @@ impl Display for Key {
     }
 }
 
+#[derive(Clone, Copy, Debug, Default, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct TimeStamp(u64);
+
+const TSO_PHYSICAL_SHIFT_BITS: u64 = 18;
+
+impl TimeStamp {
+    /// Create a time stamp from physical and logical components.
+    pub fn compose(physical: u64, logical: u64) -> TimeStamp {
+        TimeStamp((physical << TSO_PHYSICAL_SHIFT_BITS) + logical)
+    }
+
+    pub fn now_sec() -> TimeStamp {
+        TimeStamp(
+            SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .unwrap()
+                .as_secs(),
+        )
+    }
+
+    pub const fn min() -> TimeStamp {
+        TimeStamp(0)
+    }
+
+    pub const fn max() -> TimeStamp {
+        TimeStamp(std::u64::MAX)
+    }
+
+    pub const fn new(ts: u64) -> TimeStamp {
+        TimeStamp(ts)
+    }
+
+    /// Extracts physical part of a timestamp, in milliseconds.
+    pub fn extract_physical(self) -> u64 {
```

Comment:
Missing `extract_logical`?

Reply:
It didn't exist before and is never used, afaict

Reply:
`extract_physical` never exists as well

Reply:
It was here: https://github.com/tikv/tikv/pull/5857/files#diff-b6af72e31984f03f182fefddd4face63L24-L26 I think GitHub hid this diff.